Update gridsearch example for multimetric scoring #15047

pratikpc opened the pull request:
#### Reference Issues/PRs
Based on Issue #15047 proposed by @amueller 

#### What does this implement/fix? Explain your changes.
As proposed by @amueller, I added support for multimetric_scoring using `GridSearchCV` to the given example

#### Any other comments?
~~Note:-
Because [predict is only available if refit=True or I believe specified](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV.predict), I set `refit` to 'precision_score_macro'.
This is a purely personal decision and it could be changed to recall_score_macro as well.
It affects predict function for now.~~

It's set to refit=False
We calculate Best Params for a given scorer not based on the return value of the attribute, but rather by
1. Finding the Lowest Valued Index/Best Index by Test Scores for given scorer
2. Finding the Params at the Lowest Ranked Index/Best Index

Now that we have the Best Params for a scorer, we use this for:
1. Finding the SVC Estimator based on the Params
2. Using this estimator to perform predict function
3. Printing out the values

Reference Code at https://github.com/scikit-learn/scikit-learn/blob/89332d3e90e8643bef63bc05986cc922d7d718e5/sklearn/model_selection/_search.py#L726

I used the following source as a reference:- https://scikit-learn.org/stable/auto_examples/model_selection/plot_multi_metric_evaluation.html

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

pratikpc committed:
Update gridsearch example for multimetric scoring

Based on Issue #15047 proposed by @amueller
References:- https://scikit-learn.org/stable/auto_examples/model_selection/plot_multi_metric_evaluation.html

pspachtholz commented:
I think the clf should be refit on the whole training set in the for loop before predicting. Otherwise the clf determined by refit parameter is used twice.

pratikpc commented:
@pspachtholz for predict, yes.
I am yet to find a way around it.
For best_params_ I have found a way around by attempting to implement what is done behind the scenes.

My plan for predict involves using the best_params_ for the purpose on the scorer and then performing a refit on the values.
But would it be the most efficient approach @pspachtholz ?

pratikpc commented:
@pspachtholz I have followed that approach of refitting only on the best params for a given scorer in the loop and using that to perform prediction.

Could you please check it out and recommend if the approach taken by me is correct?

pspachtholz commented:
I think that looks good already.

2 small things that could be optimized in my opinion, but just my 2 cents:
- If you go for refitting in the for loop, you could also set `refit=False` in the GridsearchCV, because you're manually refitting.
- Initializing the estimator (SVC) could be factored out, e.g.:
`base_estimator = SVC()` and then in the for loop `best_estimator = base_estimator.set_params(**best_params)`
or alternatively it could also be retrieved from the Gridsearch instance `base_estimator = clf.estimator`. 

While this modifies the estimator in place, which i think is fine for the example, the cleanest way would be to clone the estimator to have a fresh copy of it.

pratikpc commented:
@pspachtholz done!
Thank you for the assistance.
I was looking for something similar to `set_params` but couldn't find it.
In the end I went with the `clone` based approach as IMHO it's the cleanest, and quite possibly the most safest method.

Could you please check it once again?
Thank you for the assistance.

pspachtholz commented:
It looks really good to me now. Maybe you could add a comment to explain the logic behind refit false and manual refitting with multiple metrics.

amueller reviewed (approved):
Looks good, thanks!

adrinjalali commented:
I think it'd be very useful as an example if we use a callable refit to choose the best model to be trained on the whole data instead of doing it manually.

amueller commented:
why a callable, we can do ``refit='recall_macro'``, right?

cmarmo commented:
Hi @pratikpc , are you still interested in working on this? You have one approval already.

pratikpc committed:
Moved refit to callable

pratikpc commented:
@adrinjalali thanks for the suggestions
I used a callable refit function

@cmarmo yeah I am still working on it

Do tell if the current one is ready

pratikpc committed:
Flake8 Formatting issues fixed

glemaitre commented:
@adrinjalali @amueller I reviewed this example and I had some concerns:

1. I find that selected from the best of the precision and recall together is kinda weird. I would expect something like breaking ties using the scoring time, for instance.
2. I find really dangerous to show how to evaluate a model within the refit on an external test set. I would make this evaluation outside using `clf.predict` using the `best_params`.

I might see the point of 2. The only benefit that I see is to be able to compute multiple score on the test set without to call the associated metric functions. I find it dangerous to show it in our example: a beginner might do weird things while an experienced user will not need this example to achieve this results.

However, I am happy to hear about thoughts leading to this example.

I proposed the following diff which make the example as a notebook style. In addition, it will select the best models in terms of precision/recall and make a tie breaking using the scoring time. The evaluation on the test set is done after the grid-search fit.

<details>

```diff
diff --git a/examples/model_selection/plot_grid_search_digits.py b/examples/model_selection/plot_grid_search_digits.py
index a873a6e319..ae95e28005 100644
--- a/examples/model_selection/plot_grid_search_digits.py
+++ b/examples/model_selection/plot_grid_search_digits.py
@@ -13,90 +13,152 @@ the model selection step.
 
 More details on tools available for model selection can be found in the
 sections on :ref:`cross_validation` and :ref:`grid_search`.
-
 """
-from sklearn import datasets
-from sklearn.base import clone
-from sklearn.model_selection import train_test_split
-from sklearn.model_selection import GridSearchCV
-from sklearn.metrics import classification_report
-from sklearn.svm import SVC
 
 print(__doc__)
 
-# Loading the Digits dataset
+# %%
+# The dataset
+# -----------
+#
+# We will work with the `digits` dataset. The goal is to classify handwritten
+# digits images.
+from sklearn import datasets
+
 digits = datasets.load_digits()
 
-# To apply an classifier on this data, we need to flatten the image, to
-# turn the data in a (samples, feature) matrix:
+# %%
+# To apply a classifier on this data, we need to flatten the images.
+# Each image of 8 by 8 pixels needs to be transformed to a vector of 64 pixels.
+# Thus, we will get a final data array of shape (n_images, n_pixels)
 n_samples = len(digits.images)
 X = digits.images.reshape((n_samples, -1))
 y = digits.target
+print(
+    f"The number of images is {X.shape[0]} and each image contains "
+    f"{X.shape[1]} pixels"
+)
 
-# Split the dataset in two equal parts
-X_train, X_test, y_train, y_test = train_test_split(
-    X, y, test_size=0.5, random_state=0)
+# %%
+# As presented in the introduction, the data will be split into a development
+# and evaluation set of equal size.
+from sklearn.model_selection import train_test_split
 
-# Set the parameters by cross-validation
-tuned_parameters = [{'kernel': ['rbf'], 'gamma': [1e-3, 1e-4],
-                     'C': [1, 10, 100, 1000]},
-                    {'kernel': ['linear'], 'C': [1, 10, 100, 1000]}]
+X_train, X_test, y_train, y_test = train_test_split(
+    X, y, test_size=0.5, random_state=0
+)
 
+# %%
+# Define our grid-search framework
+# --------------------------------
+#
+# We will fine-tune a classifier by searching the best hyper-parameters on the
+# development dataset. We need to define the values of the hyper-parameters
+# and the scores to compute to later on select the best candidate.
+tuned_parameters = [
+    {"kernel": ["rbf"], "gamma": [1e-3, 1e-4], "C": [1, 10, 100, 1000]},
+    {"kernel": ["linear"], "C": [1, 10, 100, 1000]},
+]
 scores = ['precision_macro', 'recall_macro']
 
-print("# Tuning hyper-parameters")
-print()
-
-
-# Returns the Best Index
-def refit_best_index(results):
-    # Set the Selected Index as -1
-    # As such, if we have no Scoring, then an error will be thrown
-    best_index = -1
-    # Set the Best Score as 0
-    best_score = 0
+# %%
+# We can also define a function that will implement the strategy to use to
+# select the best candidate from the `cv_results_` returns from the
+# :class:`~sklearn.model_selection.GridSearchCV`. Once the candidate selected,
+# it will be automatically refitted by the
+# :class:`~sklearn.model_selection.GridSearchCV` instance. Here, the strategy
+# is to short list the model which are the best in terms of precision and
+# recall. From the selected models, we finally select the fastest model at
+# predicting.
+#
+# Note that these choices are completely arbitrary.
+import pandas as pd
+
+
+def refit_strategy(cv_results):
+    """Define the strategy to select the best estimator.
+
+    The strategy defines here is to select the best ranking model in terms of
+    precision and recall. Once these models are selected, we can select the
+    fastest model to predict.
+
+    Parameters
+    ----------
+    cv_results : dict of numpy (masked) ndarrays
+        CV results as returned by the `GridSearchCV`.
+
+    Returns
+    -------
+    best_index : int
+        The index of the best estimator as it appears in `cv_results`.
+    """
+    # print the info about the grid-search for the different scores
     for score in scores:
-        print("Best parameters set found on development set %s:" % score)
-        print()
-        best_score_index = results["rank_test_%s" % score].argmin()
-        best_score_params = results["params"][best_score_index]
-        print()
-        print("Grid scores on development set:")
-        print()
-        means = results['mean_test_%s' % score]
-        stds = results['std_test_%s' % score]
-
-        # Check if this is the Biggest Index
-        if(best_score < means[best_score_index]):
-            best_score = means[best_score_index]
-            best_index = best_score_index
-
-        for mean, std, params in zip(means, stds, results['params']):
-            print("%0.3f (+/-%0.03f) for %r"
-                  % (mean, std * 2, params))
-        print()
-
-        print("Detailed classification report:")
-        print()
-        print("The model is trained on the full development set.")
-        print("The scores are computed on the full evaluation set.")
-        print()
-
-        best_estimator = clone(clf.estimator)
-        best_estimator.set_params(**best_score_params)
-        best_estimator.fit(X_train, y_train)
-        print("Best parameters set found on development set")
-        print(best_score_params)
-        y_true, y_pred = y_test, best_estimator.predict(X_test)
-        print(classification_report(y_true, y_pred))
-        print()
-    return best_index
-
+        print("\nGrid scores on development set using {score}:\n")
+        for mean, std, params in zip(
+            cv_results[f"mean_test_{score}"],
+            cv_results[f"std_test_{score}"],
+            cv_results["params"],
+        ):
+            print(f"{mean:0.3f} (+/-{std * 2:0.03f}) for {params}")
+
+    cv_results_ = pd.DataFrame(cv_results)
+    cv_results_ = cv_results_[
+        [
+            "mean_score_time",
+            "mean_test_recall_macro",
+            "mean_test_precision_macro",
+            "rank_test_recall_macro",
+            "rank_test_precision_macro",
+        ]
+    ]
+
+    # Select the most performant models in terms of precision/recall
+    mask_best_recall = cv_results_["rank_test_recall_macro"] == 1
+    mask_best_precision = cv_results_["rank_test_precision_macro"] == 1
+    mask_combined = mask_best_recall & mask_best_precision
+    print(
+        f"The following models are the most performance in terms of "
+        f"recall/precision:\n\n {cv_results_[mask_combined]}"
+    )
+
+    # From the best candidates, select the fastest model to predict
+    best_index = cv_results_.loc[mask_combined, "mean_score_time"].idxmin()
+
+    print(
+        f"\nWe selected the model with the following parameters which maximize"
+        f" the performance metric and minimize the scoring time:\n\n"
+        f"{cv_results_.loc[best_index]}"
+    )
+
+    return cv_results_.index[best_index]
+
+
+# %%
+# Once we defined our strategy to select the best model, we can create the
+# grid-search instance. Subsequently, we can check the best parameters found.
+from sklearn.model_selection import GridSearchCV
+from sklearn.svm import SVC
 
 clf = GridSearchCV(
-    SVC(), tuned_parameters, scoring=scores, refit=refit_best_index
+    SVC(), tuned_parameters, scoring=scores, refit=refit_strategy
 )
+
+print("# Tuning hyper-parameters \n")
 clf.fit(X_train, y_train)
+print(f"\nThe best set of parameters found are:\n{clf.best_params_}")
+
+# %%
+# Finally, we evaluate the fine-tuned model on the left-out evaluation set.
+from sklearn.metrics import classification_report
+
+y_pred = clf.predict(X_test)
+print(
+    f"\nOur selected model will have the following performance on the "
+    f"evaluation set:\n\n {classification_report(y_test, y_pred)}"
+)
 
-# Note the problem is too easy: the hyperparameter plateau is too flat and the
-# output model is the same for precision and recall with ties in quality.
+# %%
+# .. note::
+#    The problem is too easy: the hyperparameter plateau is too flat and the
+#    output model is the same for precision and recall with ties in quality.
```

pratikpc committed:
Convert to Notebook style. Select the best models in terms of precision/recall and make a tie breaking using the scoring time. The evaluation on the test set is done after the grid-search fit.

Based on https://github.com/scikit-learn/scikit-learn/pull/15054#issuecomment-712376570

pratikpc commented:
@glemaitre I quite liked your diff and I have made the requisite changes

Converting to Notebook form as it currently stands would certainly help increase readability and the comments you have added would certainly make reading and comprehending docs much more simpler

glemaitre reviewed (approved):
LGTM but then I am a bit biased.

@adrinjalali @amueller Do you agree with the reformatting. You can see the artifacts there: https://122892-843222-gh.circle-artifacts.com/0/doc/auto_examples/model_selection/plot_grid_search_digits.html

adrinjalali commented:
This looks quite good, I'm happy with it. I'd only add a link to the example from the API guide so that people can see how they can use a callable for `refit`. Otherwise LGTM.

cmarmo commented:
Hi @pratikpc , will you be able to address the last comment from @adrinjalali? The failing check will probably go if you synchronize with upstream. Finishing this by the end of the week will make this PR good for 0.24. Thanks for your work so far and for your patience!

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into pr/15054

pratikpc commented:
@cmarmo @adrinjalali a link to this page?
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html

cmarmo commented:
@adrinjalali I think the link from the API page of [GridSearchCV](https://scikit-learn.org/dev/modules/generated/sklearn.model_selection.GridSearchCV.html) to the example is already done in the [dev version of the documentation](https://scikit-learn.org/dev/modules/generated/sklearn.model_selection.GridSearchCV.html#examples-using-sklearn-model-selection-gridsearchcv) ... perhaps this PR is mergeable as is? Thanks!

adrinjalali commented:
Those links are not discoverable, and especially for a case like `GridSeachCV`, there are too many of them and the user is lost and cannot find the relevant information. I'm asking to add a link to this specific example right in the docstring of `refit`, and not on the bottom of the page along with the other examples.

cmarmo commented:
@pratikpc , do you mind adding a link to the example from the GridSearchCV documentation? For example from here https://github.com/scikit-learn/scikit-learn/blob/42aff4e2edd8e8887478f6ff1628f27de97be6a3/sklearn/model_selection/_search.py#L1073
You already have two approvals. Thanks for your work so far.

cmarmo commented:
Following the [triaging workflow](https://scikit-learn.org/dev/developers/bug_triaging.html?highlight=triaging#triaging-operations-for-members-of-the-core-and-contributor-experience-teams) I'm closing this pull request as superseded by #22279.
